File: 647-palindromic-substrings/647-palindromic-substrings.java
class Solution {
    public int countSubstrings(String s) {
        int res = 0;
        for(int i = 0; i < s.length(); i++) {
            res += panlin(s, i, i);
            res += panlin(s, i, i + 1);
        }
        return res;
    }
    public int panlin(String s, int i, int j) {
        int res = 0;
        while(i >= 0 && j < s.length() && s.charAt(i) == s.charAt(j)) {
            res++;
            i--;
            j++;
        }
        return res;
    }
}

File: 647-palindromic-substrings/647-palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n, res = len(s), 0
        dp = [[False]*len(s) for _ in range(n)]
        
        for i in range(n -1, -1, -1):
            for j in range(i, n):
                dp[i][j] = s[i] == s[j] and (j - i + 1 <=2 or dp[i + 1][j - 1])
                if dp[i][j]:
                    res += 1
        return res